### ISSUE TITLE
IOperation support for patterns

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#19927]
Proposed API:

```
    public interface IPattern : IOperation
    {
    }

    public interface IDeclarationPattern : IPattern
    {
        ISymbol DeclaredSymbol { get; }
    }

    public interface IConstantPattern : IPattern
    {
        IOperation Value { get; }
    }

    public interface IPatternCaseClause : ICaseClause
    {
        ILabelSymbol Label { get; }
        IPattern Pattern { get; }
        IOperation GuardExpression { get; }
    }

    public interface IIsPatternExpression : IOperation
    {
        IOperation Expression { get; }
        IPattern Pattern { get; }
    }
```

**Comment by jinujoseph** [comment]
**Design Team Decision**

Approves the above API proposal.

**Comment by jinujoseph** [comment]
closed via #20276
